skmorf/tree/_unsup_criterion.pyx

```diff
@@ -133,35 +160,181 @@ cdef class TwoMeans(Criterion):
+                    w = sample_weight[i]
+
+                for k in range(self.n_outputs):
+                    self.sum_left[k, <SIZE_t> self.y[i, k]] -= w
+
+                self.weighted_n_left -= w
+
+        # Update right part statistics
+        self.weighted_n_right = self.weighted_n_node_samples - self.weighted_n_left
+        for k in range(self.n_outputs):
+            for c in range(self.n_classes[k]):
+                self.sum_right[k, c] = self.sum_total[k, c] - self.sum_left[k, c]
+
+        self.pos = new_pos
+        return 0
 
     cdef double node_impurity(self) nogil:
-        # need to implement
-        pass
+        """Evaluate the impurity of the current node.
+
+        Evaluate the TwoMeans criterion as impurity of the current node,
+        i.e. the impurity of sample_indices[start:end]. The smaller the impurity the
+        better.
+        """
+        cdef SIZE_t pos = self.pos
+        cdef SIZE_t end = self.end
+        cdef const DTYPE_t[:, :] X = self.X
+
+        cdef double impurity = 0.0
+        cdef double mu_left #left mean
+        cdef double mu_right #right mean
+        cdef double left_c = 0.0 #left cluster
+        cdef double right_c = 0.0 #right cluster
+        cdef SIZE_t k
+        cdef SIZE_t c
+
+        mu_left = X[:pos].sum()
+        mu_right = X[pos:end].sum()
+
+        for k in range(pos):
+            left_s += (X[k] - mu_left)**2
+
+        for c in range(pos,end):
+            right_s += (X[c] - mu_right)**2
+
+        impurity = left_s + right_s
+        impurity /= self.weighted_n_node_samples
+
+        return impurity
 
     cdef void children_impurity(self, double* impurity_left,
                                 double* impurity_right) nogil:
-        # need to implement
+        # cdef double entropy_left = 0.0
+        # cdef double entropy_right = 0.0
+        # cdef double count_k
+        # cdef SIZE_t k
+        # cdef SIZE_t c
+
+        # for k in range(self.n_outputs):
+        #     for c in range(self.n_classes[k]):
+        #         count_k = self.sum_left[k, c]
+        #         if count_k > 0.0:
+        #             count_k /= self.weighted_n_left
+        #             entropy_left -= count_k * log(count_k)
+
+        #         count_k = self.sum_right[k, c]
+        #         if count_k > 0.0:
+        #             count_k /= self.weighted_n_right
+        #             entropy_right -= count_k * log(count_k)
+
+        # impurity_left[0] = entropy_left / self.n_outputs
+        # impurity_right[0] = entropy_right / self.n_outputs
+        
+        # use left_sum and right_sum to calculate impurity?
+        # or should this method take two separate Xs making up
+        # left and right child nodes?
         pass
 
     cdef void node_value(self, double* dest) nogil:
-        """Compute the node value of samples[start:end] and save it into dest.
+        r"""Compute the node value of samples[start:end] and save it into dest.
```

Comment:
No, it does not handle the computation. See scikit-learn splitter and tree for how `node_value()` is used.